pkg/controller/endpoint/endpoints_controller_test.go

```diff
@@ -1954,6 +1954,146 @@ func TestEndpointPortFromServicePort(t *testing.T) {
 	}
 }
 
+func TestMultipleServiceChanges(t *testing.T) {
+	mux := http.NewServeMux()
+	ns := metav1.NamespaceDefault
+	expectedSubsets := []v1.EndpointSubset{{
+		Addresses: []v1.EndpointAddress{
+			{IP: "1.2.3.4", NodeName: &emptyNodeName, TargetRef: &v1.ObjectReference{Kind: "Pod", Name: "pod0", Namespace: ns}},
+		},
+	}}
+	endpoint := &v1.Endpoints{
+		ObjectMeta: metav1.ObjectMeta{Name: "foo", Namespace: ns, ResourceVersion: "1"},
+		Subsets:    expectedSubsets,
+	}
+
+	var endpointsStore cache.Store
+	var endpoints *endpointController
+	blockDelete := make(chan struct{})
+	blockNextAction := make(chan struct{})
+	stopChan := make(chan struct{})
+
+	handlerFunc := func(res http.ResponseWriter, req *http.Request) {
+		if endpointsStore == nil {
+			res.WriteHeader(http.StatusInternalServerError)
+			res.Write([]byte("endpointsStore has not been set yet"))
+			return
+		}
+		if endpoints == nil {
+
+			res.WriteHeader(http.StatusInternalServerError)
+			res.Write([]byte("endpoints"))
+			return
+		}
+
+		if req.Method == "POST" {
+			endpointsStore.Add(endpoint)
+			blockNextAction <- struct{}{}
+		}
+
+		if req.Method == "DELETE" {
+			go func() {
+				// Delay the deletion of endoints to make endpoint cache out of sync
+				<-blockDelete
+				endpointsStore.Delete(endpoint)
+				endpoints.onEndpointsDelete(endpoint)
+			}()
+			blockNextAction <- struct{}{}
+		}
+
+		res.WriteHeader(http.StatusOK)
+		res.Write([]byte(runtime.EncodeOrDie(clientscheme.Codecs.LegacyCodec(v1.SchemeGroupVersion), &v1.Endpoints{})))
+	}
+
+	mux.HandleFunc("/api/v1/namespaces/"+ns+"/endpoints", handlerFunc)
+	mux.HandleFunc("/api/v1/namespaces/"+ns+"/endpoints/", handlerFunc)
+	mux.HandleFunc("/api/v1/namespaces/"+ns+"/events", func(res http.ResponseWriter, req *http.Request) {})
+	mux.HandleFunc("/", func(res http.ResponseWriter, req *http.Request) {
+		t.Errorf("unexpected request: %v", req.RequestURI)
+		http.Error(res, "", http.StatusNotFound)
+	})
+	testServer := httptest.NewServer(mux)
```

Comment:
Would it be possible to extract this and the handlerFunc into some kind of helper func?